File: java-service/src/main/java/com/nexchat/security/JwtService.java
/**
 * NexChat — JWT Service
 *
 * Encapsulates all JWT operations: generation, validation, and claim extraction.
 * This service is the cryptographic core of the authentication system.
 *
 * Token Lifecycle:
 *   1. Login  → generateAccessToken() + generateRefreshToken()
 *   2. Request → extractUserId() → validate() → authorize
 *   3. Refresh → validate refresh token → generateAccessToken()
 *   4. Logout  → mark session inactive in DB (refresh token revoked)
 *
 * Algorithm: HMAC-SHA256 (HS256) — symmetric, single-key signing.
 * The signing key is derived from the JWT_SECRET environment variable.
 */
package com.nexchat.security;

import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;
import io.jsonwebtoken.security.SignatureException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Date;
import java.util.UUID;

@Service
@Slf4j // Lombok: generates private static final Logger log = LoggerFactory.getLogger(JwtService.class)
public class JwtService {

    /*
     * @Value reads from application.yml's nexchat.jwt.secret.
     * The secret MUST be at least 256 bits (32 bytes) for HMAC-SHA256.
     * Our default in application.yml is intentionally long for dev safety.
     */
    @Value("${nexchat.jwt.secret}")
    private String jwtSecret;

    @Value("${nexchat.jwt.access-token-expiration}")
    private long accessTokenExpirationMs;

    @Value("${nexchat.jwt.refresh-token-expiration}")
    private long refreshTokenExpirationMs;

    /*
     * Lazily computed signing key. We derive the SecretKey from the raw secret
     * string using Keys.hmacShaKeyFor(). The key bytes must be >= 256 bits for HS256.
     *
     * ⚠️ We recompute this on every call rather than caching it as a field
     * because @Value injection happens AFTER field initialization. If we used
     * a field initializer, jwtSecret would still be null.
     */
    private SecretKey getSigningKey() {
        byte[] keyBytes = jwtSecret.getBytes(StandardCharsets.UTF_8);
        return Keys.hmacShaKeyFor(keyBytes);
    }

    /**
     * Generates a short-lived access token.
     *
     * Claims embedded in the token (readable without verification):
     * - sub (subject): the user's UUID — primary identifier
     * - type: "ACCESS" — distinguishes token types server-side
     * - iat (issued at): auto-added by JJWT
     * - exp (expiration): auto-added from expiration Date
     *
     * @param userId The authenticated user's UUID
     * @return Signed JWT string, e.g. "eyJhbGc..."
     */
    public String generateAccessToken(UUID userId) {
        Instant now = Instant.now();
        Instant expiresAt = now.plusMillis(accessTokenExpirationMs);

        return Jwts.builder()
            .subject(userId.toString())       // 'sub' claim
            .claim("type", "ACCESS")          // custom claim for token type check
            .issuedAt(Date.from(now))         // 'iat' claim
            .expiration(Date.from(expiresAt)) // 'exp' claim
            .signWith(getSigningKey())         // signs with HS256 by default
            .compact();                        // serializes to "header.payload.signature"
    }

    /**
     * Generates a long-lived refresh token.
     *
     * ⚠️ IMPORTANT: Refresh tokens are also JWTs, but we ADDITIONALLY store
     * a hash of them in the sessions table. This is what enables revocation.
     * The JWT alone can't be revoked — the DB record can.
     *
     * @param userId The authenticated user's UUID
     * @return Signed JWT string for refresh use
     */
    public String generateRefreshToken(UUID userId) {
        Instant now = Instant.now();
        Instant expiresAt = now.plusMillis(refreshTokenExpirationMs);

        return Jwts.builder()
            .subject(userId.toString())
            .claim("type", "REFRESH")
            .issuedAt(Date.from(now))
            .expiration(Date.from(expiresAt))
            .signWith(getSigningKey())
            .compact();
    }

    /**
     * Extracts the user UUID from a token's 'sub' claim.
     * This is the ONLY user identifier embedded in the token.
     *
     * @param token The raw JWT string (without "Bearer " prefix)
     * @return The user's UUID
     * @throws JwtException if token is invalid or expired
     */
    public UUID extractUserId(String token) {
        String subject = extractAllClaims(token).getSubject();
        return UUID.fromString(subject);
    }

    /**
     * Validates that a token:
     * 1. Has a valid signature (wasn't tampered with)
     * 2. Hasn't expired
     * 3. Has the expected type claim
     *
     * @param token The raw JWT string
     * @param expectedType "ACCESS" or "REFRESH"
     * @return true if valid
     */
    public boolean isTokenValid(String token, String expectedType) {
        try {
            Claims claims = extractAllClaims(token);
            String tokenType = claims.get("type", String.class);
            return expectedType.equals(tokenType);
            // If we reach here, the signature was valid and the token isn't expired.
            // (JJWT throws ExpiredJwtException automatically on expiry.)
        } catch (ExpiredJwtException e) {
            log.debug("JWT token expired: {}", e.getMessage());
            return false;
        } catch (SignatureException e) {
            // ⚠️ SECURITY: Log at WARN level — this could be an attack.
            log.warn("Invalid JWT signature detected — possible token tampering");
            return false;
        } catch (MalformedJwtException e) {
            log.debug("Malformed JWT token: {}", e.getMessage());
            return false;
        } catch (JwtException e) {
            log.debug("JWT validation failed: {}", e.getMessage());
            return false;
        }
    }

    /**
     * Returns the expiration Instant for access tokens.
     * Used to populate the 'accessTokenExpiresAt' field in AuthResponse.
     */
    public Instant getAccessTokenExpiresAt() {
        return Instant.now().plusMillis(accessTokenExpirationMs);
    }

    /*
     * Core parsing method. All extract/validate methods delegate here.
     * Throws JwtException subclasses on any validation failure.
     *
     * The JwtParserBuilder verifyWith() sets the key used to verify the signature.
     * If the signature doesn't match, JJWT throws SignatureException.
     * If the token is expired, JJWT throws ExpiredJwtException.
     */
    private Claims extractAllClaims(String token) {
        return Jwts.parser()
            .verifyWith(getSigningKey())
            .build()
            .parseSignedClaims(token)
            .getPayload();
    }
}

File: backend/src/services/jwt.service.ts
/**
 * NexChat — JWT Service
 *
 * Mirrors: com.nexchat.security.JwtService
 *
 * Algorithm : HMAC-SHA256 (HS256) — symmetric, single shared secret.
 * Claims     : sub (userId string), type ("ACCESS"|"REFRESH"), iat, exp
 *
 * Token Lifecycle:
 *   1. Login/Register → signAccessToken() + signRefreshToken()
 *   2. Every request  → verifyAccessToken() → extracts userId
 *   3. Refresh        → verifyRefreshToken() → issue new access token
 *   4. Logout         → revoke refresh token hash in sessions collection
 *
 * Security notes:
 *  - JWT secret must be ≥ 256 bits (32 bytes) for HS256 — validated in config.
 *  - Access tokens: 15 min (short-lived — limit exposure window).
 *  - Refresh tokens: 7 days (long-lived — stored hash enables revocation).
 *  - Token type claim prevents an access token being used as a refresh or vice-versa.
 *  - Expired tokens throw a distinct error (JwtTokenExpiredError) so middleware
 *    can return 401 with a specific message that clients use to trigger refresh.
 */
import jwt from "jsonwebtoken";
import { config } from "../config/index.js";
import { logger } from "../utils/logger.js";

// ── Token type discriminator (matches Java: "ACCESS" | "REFRESH") ─────────────
export const TOKEN_TYPE = {
  ACCESS: "ACCESS",
  REFRESH: "REFRESH",
} as const;
export type TokenType = (typeof TOKEN_TYPE)[keyof typeof TOKEN_TYPE];

// ── Payload structure (what we embed inside each JWT) ─────────────────────────
export interface JwtPayload {
  /** The authenticated user's UUID string — matches MongoDB _id */
  sub: string;
  /** Token type — prevents cross-purpose misuse */
  type: TokenType;
  /** Issued-at (Unix seconds) */
  iat: number;
  /** Expiration (Unix seconds) */
  exp: number;
}

// ── Custom error classes ───────────────────────────────────────────────────────

/** Token has a valid signature but is past its expiry time. */
export class JwtTokenExpiredError extends Error {
  constructor() {
    super("Token has expired");
    this.name = "JwtTokenExpiredError";
  }
}

/** Token is syntactically malformed, signature is invalid, or type is wrong. */
export class JwtTokenInvalidError extends Error {
  constructor(reason?: string) {
    super(reason ?? "Token is invalid");
    this.name = "JwtTokenInvalidError";
  }
}

// ── Token generation ──────────────────────────────────────────────────────────

/**
 * Sign a short-lived ACCESS token.
 * Default expiry: 15 minutes (configurable via JWT_ACCESS_EXPIRY_MS env).
 *
 * @param userId  The user's UUID string (stored as 'sub' claim)
 * @returns       Signed JWT string
 */
export function signAccessToken(userId: string): string {
  return jwt.sign(
    { sub: userId, type: TOKEN_TYPE.ACCESS },
    config.jwtSecret,
    {
      algorithm: "HS256",
      expiresIn: Math.floor(config.jwtAccessExpiryMs / 1000), // seconds
    }
  );
}

import crypto from "crypto";

/**
 * Sign a long-lived REFRESH token.
 * Default expiry: 7 days (configurable via JWT_REFRESH_EXPIRY_MS env).
 *
 * ⚠️ After signing, the CALLER MUST store a SHA-256 hash of this token
 * in the sessions collection to enable revocation (session.repository.ts).
 *
 * @param userId  The user's UUID string
 * @returns       Signed JWT string
 */
export function signRefreshToken(userId: string): string {
  return jwt.sign(
    { sub: userId, type: TOKEN_TYPE.REFRESH, jti: crypto.randomUUID() },
    config.jwtSecret,
    {
      algorithm: "HS256",
      expiresIn: Math.floor(config.jwtRefreshExpiryMs / 1000),
    }
  );
}

/**
 * Return the Date when a newly-minted access token will expire.
 * Used to populate 'accessTokenExpiresAt' in the AuthResponse — matches
 * JwtService.getAccessTokenExpiresAt() in Java.
 */
export function getAccessTokenExpiresAt(): Date {
  return new Date(Date.now() + config.jwtAccessExpiryMs);
}

// ── Token verification ────────────────────────────────────────────────────────

/**
 * Verify and decode a token.
 * Throws JwtTokenExpiredError or JwtTokenInvalidError — never returns null.
 *
 * @internal — use verifyAccessToken / verifyRefreshToken for type safety.
 */
function verifyToken(raw: string, expectedType: TokenType): JwtPayload {
  let payload: JwtPayload;

  try {
    payload = jwt.verify(raw, config.jwtSecret, {
      algorithms: ["HS256"],
    }) as JwtPayload;
  } catch (err) {
    if (err instanceof jwt.TokenExpiredError) {
      throw new JwtTokenExpiredError();
    }
    // NotBeforeError, JsonWebTokenError, SyntaxError, etc.
    logger.warn({ err }, "JWT verification failed");
    throw new JwtTokenInvalidError(
      err instanceof Error ? err.message : "Malformed token"
    );
  }

  // Verify the type claim — prevent access tokens being used for refresh
  if (payload.type !== expectedType) {
    throw new JwtTokenInvalidError(
      `Expected ${expectedType} token, got ${payload.type}`
    );
  }

  return payload;
}

/**
 * Verify an ACCESS token and extract its payload.
 *
 * @throws JwtTokenExpiredError  — token is past expiry (client should refresh)
 * @throws JwtTokenInvalidError  — signature is bad, token is malformed, etc.
 */
export function verifyAccessToken(raw: string): JwtPayload {
  return verifyToken(raw, TOKEN_TYPE.ACCESS);
}

/**
 * Verify a REFRESH token and extract its payload.
 * The caller MUST also check the sessions collection — this only verifies
 * the cryptographic signature, not whether the session is active/revoked.
 *
 * @throws JwtTokenExpiredError
 * @throws JwtTokenInvalidError
 */
export function verifyRefreshToken(raw: string): JwtPayload {
  return verifyToken(raw, TOKEN_TYPE.REFRESH);
}

/**
 * Decode a token WITHOUT verifying the signature.
 * Use only when you need the claims after expiry (e.g., to extract userId
 * for logging). Never use for authentication decisions.
 */
export function decodeTokenUnsafe(raw: string): JwtPayload | null {
  try {
    return jwt.decode(raw) as JwtPayload | null;
  } catch {
    return null;
  }
}
